File: engine/dynamic/render_vulkan/src/util/image.cpp
/*
 * This file is a part of Argus.
 * Copyright (c) 2019-2024, Max Roncace <mproncace@protonmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "argus/lowlevel/math.hpp"

#include "argus/core/engine.hpp"

#include "internal/render_vulkan/setup/device.hpp"
#include "internal/render_vulkan/util/image.hpp"
#include "internal/render_vulkan/util/memory.hpp"

#include "vulkan/vulkan.h"
#include "internal/render_vulkan/util/command_buffer.hpp"

#include <vector>

#include <cstdint>

namespace argus {
    VkImage create_image(const LogicalDevice &device, VkFormat format, const Vector2u &size,
            VkImageUsageFlags usage) {
        VkExtent3D extent = { size.x, size.y, 1 };
        std::vector<uint32_t> qf_indices { device.queue_indices.graphics_family };

        VkImageCreateInfo image_info {};
        image_info.sType = VK_STRUCTURE_TYPE_IMAGE_CREATE_INFO;
        image_info.imageType = VK_IMAGE_TYPE_2D;
        image_info.format = format;
        image_info.extent = extent;
        image_info.mipLevels = 1;
        image_info.arrayLayers = 1;
        image_info.samples = VK_SAMPLE_COUNT_1_BIT;
        image_info.tiling = VK_IMAGE_TILING_OPTIMAL;
        image_info.usage = usage;
        image_info.sharingMode = VK_SHARING_MODE_EXCLUSIVE;
        image_info.queueFamilyIndexCount = static_cast<uint32_t>(qf_indices.size());
        image_info.pQueueFamilyIndices = qf_indices.data();
        image_info.initialLayout = VK_IMAGE_LAYOUT_UNDEFINED;

        VkImage image;
        {
            std::lock_guard<std::mutex> queue_lock(device.queue_mutexes->graphics_family);
            vkCreateImage(device.logical_device, &image_info, nullptr, &image);
        }

        VkMemoryRequirements mem_reqs;
        vkGetImageMemoryRequirements(device.logical_device, image, &mem_reqs);

        VkMemoryAllocateInfo alloc_info = {};
        alloc_info.sType = VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_INFO;
        alloc_info.allocationSize = mem_reqs.size;
        alloc_info.memoryTypeIndex = find_memory_type(device, mem_reqs.memoryTypeBits,
                GraphicsMemoryPropCombos::DeviceRo);

        VkDeviceMemory image_memory;
        if (vkAllocateMemory(device.logical_device, &alloc_info, nullptr, &image_memory) != VK_SUCCESS) {
            crash("Failed to allocate memory for image");
        }

        if (vkBindImageMemory(device.logical_device, image, image_memory, 0) != VK_SUCCESS) {
            crash("Failed to bind image memory");
        }

        return image;
    }

    VkImageView create_image_view(const LogicalDevice &device, VkImage image, VkFormat format,
            VkImageAspectFlags aspect_mask) {
        VkImageViewCreateInfo view_info {};
        view_info.sType = VK_STRUCTURE_TYPE_IMAGE_VIEW_CREATE_INFO;
        view_info.image = image;
        view_info.viewType = VK_IMAGE_VIEW_TYPE_2D;
        view_info.format = format;
        view_info.components.r = VK_COMPONENT_SWIZZLE_IDENTITY;
        view_info.components.g = VK_COMPONENT_SWIZZLE_IDENTITY;
        view_info.components.b = VK_COMPONENT_SWIZZLE_IDENTITY;
        view_info.components.a = VK_COMPONENT_SWIZZLE_IDENTITY;
        view_info.subresourceRange.aspectMask = aspect_mask;
        view_info.subresourceRange.baseMipLevel = 0;
        view_info.subresourceRange.levelCount = 1;
        view_info.subresourceRange.baseArrayLayer = 0;
        view_info.subresourceRange.layerCount = 1;

        VkImageView view;
        vkCreateImageView(device.logical_device, &view_info, nullptr, &view);

        return view;
    }

    ImageInfo create_image_and_image_view(const LogicalDevice &device, VkFormat format, const Vector2u &size,
            VkImageUsageFlags usage, VkImageAspectFlags aspect_mask) {
        auto image = create_image(device, format, size, usage);
        auto view = create_image_view(device, image, format, aspect_mask);
        return ImageInfo { size, format, image, view };
    }

    void destroy_image(const LogicalDevice &device, VkImage image) {
        vkDestroyImage(device.logical_device, image, nullptr);
    }

    void destroy_image_view(const LogicalDevice &device, VkImageView view) {
        vkDestroyImageView(device.logical_device, view, nullptr);
    }

    void destroy_image_and_image_view(const LogicalDevice &device, const ImageInfo &image) {
        destroy_image_view(device, image.view);
        destroy_image(device, image.handle);
    }

    void perform_image_transition(const CommandBufferInfo &cmd_buf, VkImage image,
            VkImageLayout old_layout, VkImageLayout new_layout,
            VkAccessFlags src_access, VkAccessFlags dst_access,
            VkPipelineStageFlags src_stage, VkPipelineStageFlags dst_stage) {
        VkImageMemoryBarrier barrier {};
        barrier.sType = VK_STRUCTURE_TYPE_IMAGE_MEMORY_BARRIER;
        barrier.oldLayout = old_layout;
        barrier.newLayout = new_layout;
        barrier.srcQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
        barrier.dstQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
        barrier.srcAccessMask = src_access;
        barrier.dstAccessMask = dst_access;
        barrier.image = image;
        barrier.subresourceRange.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT;
        barrier.subresourceRange.baseMipLevel = 0;
        barrier.subresourceRange.levelCount = 1;
        barrier.subresourceRange.baseArrayLayer = 0;
        barrier.subresourceRange.layerCount = 1;

        vkCmdPipelineBarrier(cmd_buf.handle, src_stage, dst_stage, 0, 0, nullptr, 0, nullptr, 1, &barrier);
    }

    void perform_image_transition(const CommandBufferInfo &cmd_buf, const ImageInfo &image,
            VkImageLayout old_layout, VkImageLayout new_layout,
            VkAccessFlags src_access, VkAccessFlags dst_access,
            VkPipelineStageFlags src_stage, VkPipelineStageFlags dst_stage) {
        perform_image_transition(cmd_buf, image.handle, old_layout, new_layout,
                src_access, dst_access, src_stage, dst_stage);
    }
}

File: engine/modules/render_vulkan/src/util/image.rs
use ash::vk;
use argus_util::math::Vector2u;
use crate::setup::device::VulkanDevice;
use crate::setup::instance::VulkanInstance;
use crate::util::{find_memory_type, CommandBufferInfo, MEM_CLASS_DEVICE_RO};

pub(crate) struct VulkanImage {
    size: Vector2u,
    format: vk::Format,
    vk_image: vk::Image,
    vk_view: vk::ImageView,
    is_destroyed: bool,
}

impl VulkanImage {
    pub(crate) fn create_image_with_view(
        instance: &VulkanInstance,
        device: &VulkanDevice,
        format: vk::Format,
        size: Vector2u,
        usage: vk::ImageUsageFlags,
        aspect_mask: vk::ImageAspectFlags
    ) -> Result<VulkanImage, String> {
        unsafe {
            let vk_image = create_vk_image(instance, device, format, &size, usage)?;
            let vk_view = create_vk_image_view(device, vk_image, format, aspect_mask)?;
            let image_info = VulkanImage {
                size,
                format,
                vk_image,
                vk_view,
                is_destroyed: false,
            };
            Ok(image_info)
        }
    }
    
    pub(crate) fn get_size(&self) -> &Vector2u {
        &self.size
    }

    pub(crate) unsafe fn get_vk_image(&self) -> vk::Image {
        self.vk_image
    }

    pub(crate) unsafe fn get_vk_image_view(&self) -> vk::ImageView {
        self.vk_view
    }

    pub(crate) fn destroy(mut self, device: &VulkanDevice) {
        unsafe {
            destroy_vk_image_view(device, self.vk_view);
            destroy_vk_image(device, self.vk_image);
        }
        self.is_destroyed = true;
    }

    pub(crate) fn perform_transition(
        &self,
        device: &VulkanDevice,
        cmd_buf: &CommandBufferInfo,
        old_layout: vk::ImageLayout,
        new_layout: vk::ImageLayout,
        src_access: vk::AccessFlags,
        dst_access: vk::AccessFlags,
        src_stage: vk::PipelineStageFlags,
        dst_stage: vk::PipelineStageFlags,
    ) {
        unsafe {
            perform_vk_image_transition(
                device,
                self.vk_image,
                cmd_buf,
                old_layout,
                new_layout,
                src_access,
                dst_access,
                src_stage,
                dst_stage
            );
        }
    }
}

impl Drop for VulkanImage {
    fn drop(&mut self) {
        if cfg!(debug_assertions) && !self.is_destroyed {
            panic!("Vulkan image was dropped without being destroyed!")
        }
    }
}

pub(crate) unsafe fn create_vk_image(
    instance: &VulkanInstance,
    device: &VulkanDevice,
    format: vk::Format,
    size: &Vector2u,
    usage: vk::ImageUsageFlags
) -> Result<vk::Image, String> {
    let extent = vk::Extent3D { width: size.x, height: size.y, depth: 1 };
    let qf_indices = vec![device.queue_indices.graphics_family];

    let image_info = vk::ImageCreateInfo::default()
        .image_type(vk::ImageType::TYPE_2D)
        .format(format)
        .extent(extent)
        .mip_levels(1)
        .array_layers(1)
        .samples(vk::SampleCountFlags::TYPE_1)
        .tiling(vk::ImageTiling::OPTIMAL)
        .usage(usage)
        .sharing_mode(vk::SharingMode::EXCLUSIVE)
        .queue_family_indices(&qf_indices)
        .initial_layout(vk::ImageLayout::UNDEFINED);

    let image = {
        let _queue_lock = device.queue_mutexes.graphics_family.lock().unwrap();
        device.logical_device.create_image(&image_info, None)
            .map_err(|err| err.to_string())?
    };

    let mem_reqs = device.logical_device.get_image_memory_requirements(image);

    let Some(mem_type) = find_memory_type(
        instance,
        device,
        mem_reqs.memory_type_bits,
        mem_reqs.size,
        MEM_CLASS_DEVICE_RO
    )
    else {
        return Err("Failed to find suitable memory type for vkImage".to_owned());
    };
    let alloc_info = vk::MemoryAllocateInfo::default()
        .allocation_size(mem_reqs.size)
        .memory_type_index(mem_type);

    let image_memory = device.logical_device.allocate_memory(&alloc_info, None)
        .map_err(|err| err.to_string())?;

    device.logical_device.bind_image_memory(image, image_memory, 0)
        .map_err(|err| err.to_string())?;

    Ok(image)
}

pub(crate) unsafe fn create_vk_image_view(
    device: &VulkanDevice,
    image: vk::Image,
    format: vk::Format,
    aspect_mask: vk::ImageAspectFlags
) -> Result<vk::ImageView, String> {
    let components = vk::ComponentMapping::default()
        .r(vk::ComponentSwizzle::IDENTITY)
        .g(vk::ComponentSwizzle::IDENTITY)
        .b(vk::ComponentSwizzle::IDENTITY)
        .a(vk::ComponentSwizzle::IDENTITY);
    let sr_range = vk::ImageSubresourceRange::default()
        .aspect_mask(aspect_mask)
        .base_mip_level(0)
        .level_count(1)
        .base_array_layer(0)
        .layer_count(1);

    let view_info = vk::ImageViewCreateInfo::default()
        .image(image)
        .view_type(vk::ImageViewType::TYPE_2D)
        .format(format)
        .components(components)
        .subresource_range(sr_range);

    let view = device.logical_device.create_image_view(&view_info, None)
        .map_err(|err| err.to_string())?;
    Ok(view)
}

pub(crate) unsafe fn destroy_vk_image(device: &VulkanDevice, image: vk::Image) {
    device.logical_device.destroy_image(image, None);
}

pub(crate) unsafe fn destroy_vk_image_view(device: &VulkanDevice, view: vk::ImageView) {
    unsafe { device.logical_device.destroy_image_view(view, None); }
}

unsafe fn perform_vk_image_transition(
    device: &VulkanDevice,
    image: vk::Image,
    cmd_buf: &CommandBufferInfo,
    old_layout: vk::ImageLayout,
    new_layout: vk::ImageLayout,
    src_access: vk::AccessFlags,
    dst_access: vk::AccessFlags,
    src_stage: vk::PipelineStageFlags,
    dst_stage: vk::PipelineStageFlags,
) {
    let sr_range = vk::ImageSubresourceRange::default()
        .aspect_mask(vk::ImageAspectFlags::COLOR)
        .base_mip_level(0)
        .level_count(1)
        .base_array_layer(0)
        .layer_count(1);

    let barrier = vk::ImageMemoryBarrier::default()
        .old_layout(old_layout)
        .new_layout(new_layout)
        .src_queue_family_index(vk::QUEUE_FAMILY_IGNORED)
        .dst_queue_family_index(vk::QUEUE_FAMILY_IGNORED)
        .src_access_mask(src_access)
        .dst_access_mask(dst_access)
        .image(image)
        .subresource_range(sr_range);

    device.logical_device.cmd_pipeline_barrier(
        cmd_buf.get_handle(),
        src_stage,
        dst_stage,
        vk::DependencyFlags::empty(),
        &[],
        &[],
        &[barrier],
    );
}
